File: clients/blazor-admin/LanguageDisplay.cs
using Blazor.Admin.Services;

namespace Blazor.Admin;

/// <summary>
/// Regional indicator flag emoji from BCP 47 locale (e.g. tr-TR → 🇹🇷, en-US → 🇺🇸).
/// </summary>
public static class LanguageDisplay
{
    private const string Globe = "\uD83C\uDF10";

    public static string FlagEmojiForLocale(string locale)
    {
        if (string.IsNullOrWhiteSpace(locale))
        {
            return Globe;
        }

        var trimmed = locale.Trim();
        var parts = trimmed.Split(new[] { '-', '_' }, StringSplitOptions.RemoveEmptyEntries);
        string? region = null;
        if (parts.Length >= 2 && parts[1].Length == 2)
        {
            region = parts[1].ToUpperInvariant();
        }
        else if (parts.Length >= 1 && parts[0].Length == 2)
        {
            region = parts[0].ToUpperInvariant();
        }
        else if (parts.Length >= 1)
        {
            region = RegionFromLanguageCode(parts[0].ToLowerInvariant());
        }

        // "EN" is not assigned in ISO 3166-1; regional-indicator pair renders as tofu, not a flag
        if (string.Equals(region, "EN", StringComparison.Ordinal))
        {
            region = "US";
        }

        if (string.IsNullOrEmpty(region) || region.Length != 2
            || !char.IsAsciiLetter(region[0]) || !char.IsAsciiLetter(region[1]))
        {
            return Globe;
        }

        static string RegionalIndicatorPair(string r) =>
            char.ConvertFromUtf32(0x1F1E6 + char.ToUpperInvariant(r[0]) - 'A')
            + char.ConvertFromUtf32(0x1F1E6 + char.ToUpperInvariant(r[1]) - 'A');

        return RegionalIndicatorPair(region);
    }

    private static string? RegionFromLanguageCode(string lang) => lang switch
    {
        "tr" => "TR",
        "en" => "US",
        "de" => "DE",
        "fr" => "FR",
        "ar" => "SA",
        "ru" => "RU",
        "es" => "ES",
        "it" => "IT",
        "nl" => "NL",
        "pl" => "PL",
        "pt" => "PT",
        "ja" => "JP",
        "ko" => "KR",
        "zh" => "CN",
        _ => null
    };

    public static string LabelForLocale(string code) => code switch
    {
        "tr-TR" => "Türkçe",
        "en-US" => "English (US)",
        "en-GB" => "English (UK)",
        "de-DE" => "Deutsch",
        "fr-FR" => "Français",
        _ => code
    };

    public sealed record LanguageOption(string Code, string Label, string Flag);

    public static List<LanguageOption> ToLanguageOptions(IEnumerable<string> codes) =>
        codes.Select(c =>
        {
            var n = UiLanguageCode.Normalize(c);
            return new LanguageOption(n, LabelForLocale(n), FlagEmojiForLocale(n));
        }).ToList();
}

File: clients/react-admin/src/core/language-display.ts
import { normalizeUiLanguage } from "./ui-language-code";

/**
 * Regional indicator flag emoji from BCP 47 locale (e.g. tr-TR → 🇹🇷, en-US → 🇺🇸).
 */
export function flagEmojiForLocale(locale: string): string {
  const trimmed = locale.trim();
  if (!trimmed) {
    return "\uD83C\uDF10";
  }
  const parts = trimmed.split(/[-_]/);
  let region: string | null = null;
  if (parts.length >= 2 && parts[1].length === 2) {
    region = parts[1].toUpperCase();
  } else if (parts[0].length === 2) {
    region = parts[0].toUpperCase();
  } else {
    region = regionFromLanguageCode(parts[0]?.toLowerCase() ?? "");
  }
  if (region === "EN") {
    region = "US";
  }
  if (!region || !/^[A-Z]{2}$/.test(region)) {
    return "\uD83C\uDF10";
  }
  const codePoint = (letter: string) => 0x1f1e6 + letter.charCodeAt(0) - 0x41;
  return String.fromCodePoint(codePoint(region[0]), codePoint(region[1]));
}

function regionFromLanguageCode(lang: string): string | null {
  const map: Record<string, string> = {
    tr: "TR",
    en: "US",
    de: "DE",
    fr: "FR",
    ar: "SA",
    ru: "RU",
    es: "ES",
    it: "IT",
    nl: "NL",
    pl: "PL",
    pt: "PT",
    ja: "JP",
    ko: "KR",
    zh: "CN"
  };
  return map[lang] ?? null;
}

export function labelForLocale(code: string): string {
  const labels: Record<string, string> = {
    "tr-TR": "Türkçe",
    "en-US": "English (US)",
    "en-GB": "English (UK)",
    "de-DE": "Deutsch",
    "fr-FR": "Français"
  };
  return labels[code] ?? code;
}

export type LanguageOption = {
  code: string;
  label: string;
  flag: string;
};

export function toLanguageOptions(codes: string[]): LanguageOption[] {
  return codes.map((code) => {
    const normalized = normalizeUiLanguage(code);
    return {
      code: normalized,
      label: labelForLocale(normalized),
      flag: flagEmojiForLocale(normalized)
    };
  });
}
